lib/bamboo/adapters/mailgun_adapter.ex

```diff
@@ -0,0 +1,125 @@
+defmodule Bamboo.MailgunAdapter do
+  @moduledoc """
+  Sends email using Mailgun's API.
+
+  Use this adapter to send emails through Mailgun's API. Requires that an API
+  key and a domain are set in the config.
+
+  ## Example config
+
+      # In config/config.exs, or config.prod.exs, etc.
+      config :my_app, MyApp.Mailer,
+        adapter: Bamboo.MailgunAdapter,
+        api_key: "my_api_key",
+        domain: "your.domain"
+
+      # Define a Mailer. Maybe in lib/my_app/mailer.ex
+      defmodule MyApp.Mailer do
+        use Bamboo.Mailer, otp_app: :my_app
+      end
+  """
+
+  @base_uri "https://api.mailgun.net/v3/"
+  @behaviour Bamboo.Adapter
+
+  alias Bamboo.Email
+
+  defmodule ApiError do
+    defexception [:message]
+
+    def exception(%{params: params, response: response}) do
+      message = """
```

Comment:
Oh that makes perfect sense! Thanks for the example output

Sent from my iPhone

> On Apr 18, 2016, at 8:22 PM, Po Chen <notifications@github.com> wrote:
> 
> In lib/bamboo/adapters/mailgun_adapter.ex:
> 
> > +      # Define a Mailer. Maybe in lib/my_app/mailer.ex
> > +      defmodule MyApp.Mailer do
> > +        use Bamboo.Mailer, otp_app: :my_app
> > +      end
> > +  """
> > +
> > +  @base_uri "https://api.mailgun.net/v3/"
> > +  @behaviour Bamboo.Adapter
> > +
> > +  alias Bamboo.Email
> > +
> > +  defmodule ApiError do
> > +    defexception [:message]
> > +
> > +    def exception(%{params: params, response: response}) do
> > +      message = """
> @paulcsmith Mailgun authentication token is actually in the request headers.
> This error message logs the params that are sent to the endpoint plus the response, hence the auth header is not disclosed.
> 
> below is an intentionally failed test case. you can see that the auth token is no where to be found.
> 
>   2) test deliver/2 sends from, html and text body, subject, and headers (Bamboo.MailgunAdapterTest)
>      test/lib/bamboo/adapters/mailgun_adapter_test.exs:83
>      ** (Bamboo.MailgunAdapter.ApiError) There was a problem sending the email through the Mailgun API.
> 
>      Here is the response:
> 
>      %HTTPoison.Response{body: "Error!!", headers: [{"server", "Cowboy"}, {"date", "Tue, 19 Apr 2016 00:20:00 GMT"}, {"content-length", "7"}, {"cache-control", "max-age=0, private, must-revalidate"}], status_code: 500}
> 
> 
>      Here are the params we sent:
> 
>      "from=from%40foo.com&html=HTML+BODY&subject=My+Subject&text=TEXT+BODY"
> 
>      stacktrace:
>        (bamboo) lib/bamboo/adapters/mailgun_adapter.ex:52: Bamboo.MailgunAdapter.deliver/2
>        test/lib/bamboo/adapters/mailgun_adapter_test.exs:91
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>